icu4c/source/i18n/dtitvfmt.cpp

```diff
@@ -1023,29 +1026,23 @@ DateIntervalFormat::normalizeHourMetacharacters(const UnicodeString& skeleton) c
         }
         
         if (hourChar == CAP_H || hourChar == LOW_K) {
-            result.replace(metacharStart, metacharCount, hourChar);
+            result.replace(hourFieldStart, hourFieldLength, hourChar);
+            result.remove(dayPeriodStart, dayPeriodLength);
         } else {
             UnicodeString hourAndDayPeriod(hourChar);
-            switch (metacharCount) {
-                case 1:
-                case 2:
-                default:
-                    hourAndDayPeriod.append(UnicodeString(dayPeriodChar));
-                    break;
-                case 3:
-                case 4:
-                    for (int32_t i = 0; i < 4; i++) {
-                        hourAndDayPeriod.append(dayPeriodChar);
-                    }
-                    break;
-                case 5:
-                case 6:
-                    for (int32_t i = 0; i < 5; i++) {
-                        hourAndDayPeriod.append(dayPeriodChar);
-                    }
-                    break;
+            int32_t newDayPeriodLength = 0;
+            if (dayPeriodLength >= 5 || hourFieldLength >= 5) {
+                newDayPeriodLength = 5;
+            } else if (dayPeriodLength >= 3 || hourFieldLength >= 3) {
+                newDayPeriodLength = 3;
+            } else {
+                newDayPeriodLength = 1;
+            }
+            for (int32_t i = 0; i < newDayPeriodLength; i++) {
+                hourAndDayPeriod.append(dayPeriodChar);
             }
-            result.replace(metacharStart, metacharCount, hourAndDayPeriod);
+            result.replace(hourFieldStart, hourFieldLength, hourAndDayPeriod);
```

Comment:
Doesn't the replace above here potentially change the length of result? So is the following remove valid, since dayPeriodStart might be after the section that was replaced above?